### PR TITLE
Alternate sharding

### PR DESCRIPTION
**Pull request by clee2000**
Changes sharding to attempt to put all serial tests on as few shards as possible.  Parallel tests are then distributed across all shards, with most of which likely ending up on the non serial shards

Example: 8 minutes of serial tests, 20 minutes of parallel tests, 2 proc per machine, 6 machines
-> 8 + 20/2 = 18 total minutes of tests
-> 18 / 6 machines = 3 min per machine
-> all serial tests should fit on 3 machines (3min, 3 min, 2min)
-> majority of parallel tests should go on last 4 machines, one of which is shared with the serial tests

Move serial tests to run first

If I want to move to a purely numbers based sharding, this ensures that parallel tests are run with parallel tests as much as possible instead of interleaving serial + parallel tests, which decreases effectiveness of parallelization, while also ensuring that test reordering is still mostly effective.

See https://hud.pytorch.org/pytorch/pytorch/commit/73e816ee809ad96fda3ee0013ee96273a44e84d4 for example logs

### CODE REVIEW

File: tools/testing/test_selections.py [hunk 1]
```diff
@@ -67,97 +67,182 @@ def get_with_pytest_shard(
...
     known_tests = get_with_pytest_shard(known_tests, test_file_times, test_class_times)
 
     if sort_by_time:
         known_tests = sorted(known_tests, key=lambda j: j.get_time(), reverse=True)
 
-    sharded_jobs: List[ShardJob] = [ShardJob() for _ in range(num_shards)]
-    for test in known_tests:
-        if must_serial(test.name):
-            min_sharded_job = min(sharded_jobs, key=lambda j: j.get_total_time())
+    def _shard_serial(tests: List[ShardedTest], sharded_jobs: List[ShardJob]) -> None:
```
Comment (huydhn) [inline tools/testing/test_selections.py:157]: Nits:  Do you think it's useful to add boolean field in ShardedTest to indicate if it's a serial or parallel shard?  That could be print to the log as part of the shard name for debugging?

File: tools/testing/test_selections.py [hunk 2]
```diff
@@ -67,97 +67,182 @@ def get_with_pytest_shard(
...
+        for test in serial_tests
+    )
+    parallel_time = sum(
+        get_duration(test, test_file_times, test_class_times) or 0
+        for test in parallel_tests
+    )
+    total_time = serial_time + parallel_time / NUM_PROCS_FOR_SHARDING_CALC
+    estimated_time_per_shard = total_time / num_shards
+    estimated_time_limit = 0.0
+    if estimated_time_per_shard != 0:
```
Comment (huydhn) [inline tools/testing/test_selections.py:218]: Appreciate if some comment could be added here to understand how the estimated time limit is calculated

File: tools/testing/test_selections.py [hunk 3]
```diff
@@ -67,97 +67,182 @@ def get_with_pytest_shard(
...
+    shard(
+        sharded_jobs[:num_serial_shards],
+        serial_tests,
+        test_file_times,
+        test_class_times,
+        estimated_time_limit,
+        sort_by_time,
+        True,
+    )
+    shard(
```
Comment (huydhn) [inline tools/testing/test_selections.py:237]: I need to double check my python knowledge, does this even work correctly when `estimated_time_limit` is skipped in the middle like this without named parameters?

Comment (clee2000) [reply tools/testing/test_selections.py:237]: Good catch, changed the function call to named parameters

File: tools/testing/test_selections.py (hunk 1 above)
Comment (clee2000) [reply tools/testing/test_selections.py:157]: if a test gets run on the shard, it'll be printed out later whether it's serial or parallel, and I'd rather not clog up the logs with all this information

**Review by huydhn — APPROVED** [review]

**Pull request closed without merging**